agent/agentpb/subscribe.proto

```diff
@@ -0,0 +1,185 @@
+message SubscribeRequest {
+    // Topic identifies the set of events the subscriber is interested in.
+    Topic Topic = 1;
+
+    // Key is a topic-specific identifier that restricts the scope of the
+    // subscription to only events pertaining to that identifier. For example
+    // for watching the instances of a single service, the service's name is
+    // specified as the key. An empty key indicates that all events in the topic
+    // are of interest.
+    string Key = 2;
+
+    // Token is the ACL token to authenticate the request. The token must have
+    // sufficient priviledges to read the requested information otherwise events
+    // will be filtered, possibly resulting in an empty snapshot and no further
+    // updates sent.
+    string Token = 3;
+
+    // Index is the raft index the subscriber has already observed up to. This
+    // is zero on an initial streaming call, but then can be provided by a
+    // client on subsequent reconnections such that the full snapshot doesn't
+    // need to be resent if the client is up to date.
+    uint64 Index = 4;
+
+    // Filter optionally specified a bexpr filter string to apply to the event
+    // objects before returning them. This is an optimization only and the
+    // server MAY NOT filter the results. In general it may not be possible for
+    // the server to correctly apply filters without keeping a history of every
+    // object sent on the subscription which would be prohibitively costly. So
+    // in general servers DO NOT filter streaming updates and rely on the client
+    // to apply the filter before returning to the user. The server MAY use this
+    // as a hint when shipping the initial snapshot of the result to reduce data
+    // sent unnecessarily over the wire. Currently for health endpoint the
+    // server ignores this entirely but we may change that later.
+    string Filter = 5;
+
+    // Datacenter specifies the Consul datacenter the request is targetted at.
+    // If it's not the local DC (or empty) the server will forward the stream to
+    // the remote DC and proxy the results back.
+    string Datacenter = 6;
+}
+
+// Event describes a streaming update on a subscription. Events are used both to
+// describe the current "snapshot" of the result as well as ongoing mutations to
+// that snapshot.
+message Event {
+    // Topic the event was published to
+    Topic Topic = 1;
+
+    // Key is the logical identifier for the entity that was mutated.
+    string Key = 2;
+
+    // Index is the raft index at which the mutation took place. At the top
+    // level of a subscription there will always be at most one Event per index.
+    // If multiple events are published to the same topic in a single raft
+    // transaction then the batch of events will be encoded inside a single
+    // top-level event to ensure they are delivered atomically to clients.
+    uint64 Index = 3;
+
```

Comment:
Where did field ID `4` go?

Reply:
This has been refactored a few times. I could reset field numbers to be clean since it's new proto but Im not sure if it matters? Eventually fields will evolve. Do you think it's worth starting with them sequential?